install/uninstall instructions (#22)
percona/everest-doc: docs/install/installEverest.md

# Install Everest

## Setup

Percona Everest has two primary components:

- [Percona CLI](https://github.com/percona/percona-everest-cli), which connects Percona Everest to your Kubernetes cluster.
- Everest application (UI for simplicity), which you can spin up using the instructions below.

### Supported operators

- Percona Operator for MySQL Based on Percona XtraDB Cluster (PXC) 1.12

> - Percona Operator for MongoDB (PSMDB) 1.14

Reviewer: ```suggestion
- Percona Operator for MongoDB (PSMDB) 1.14.0
```

> - Percona Operator for PostgreSQL (PG) 2.2

Reviewer: ```suggestion
- Percona Operator for PostgreSQL (PG) 2.2.0
```

## Before you start

Percona Everest assists with installing all the necessary operators and required packages, but does not currently help with spinning up a Kubernetes cluster.
Therefore, before installing Everest, ensure that you have a publicly accessible Kubernetes cluster.

We recommend [setting one up on the Amazon Elastic Kubernetes Service (EKS)](../install/eks.md) or on [Google Kubernetes Engine (GKE)](../install/GKE-k8s-cluster), as it may be difficult to currently spin one up on local Kubernetes installations (minikube, kind, k3d, or similar products).

## Prerequisites

Before getting started with Percona Everest, we recommend that you:

> 1. Install [Docker](https://docs.docker.com/engine/install/) and its compose plugin (included by default in the aforementioned docker engine installation instructions).

Reviewer 2: (included by default in the aforementioned docker engine installation instructions). -> Where do we have this instruction?

2. Install [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) for EKS or the [gcloud CLI](https://cloud.google.com/sdk/docs/install) for GKE.

> 3. Download the kubeconfig file from the Kubernetes cluster that you want to use with Everest. By default, Everest uses the file available under *~/.kube/config.*

Reviewer: I am not sure about `Download the kubeconfig` because if you create k8s cluster from your local machine, this file will be created automatically.  It is really tricky point because depending on your k8s cluster (EKS/GKE) you need to have additional tool/dependency for it or some additional commands. I suggest to add something like 'You need to have access to k8s cluster' and adding some command on how the user can verify it e.g.
```
❯ kubectl get nodes
NAME                                       STATUS   ROLES    AGE   VERSION
gke-<name>-default-pool-75d48bfc-bx8g   Ready    <none>   11h   v1.26.7-gke.500
gke-<name>-default-pool-75d48bfc-c2df   Ready    <none>   11h   v1.26.7-gke.500
gke-<name>-default-pool-75d48bfc-zl7k   Ready    <none>   11h   v1.26.7-gke.500
```

## Get started

To install and provision Percona Everest:

1. Download the Docker compose file:
```

> curl https://raw.githubusercontent.com/percona/percona-everest-backend/v0.3.0/deploy/quickstart-compose.yml -o quickstart-compose.yml

Reviewer 3: Correct path: curl https://raw.githubusercontent.com/percona/percona-everest-backend/release-0.3.0/deploy/quickstart-compose.yml -o quickstart-compose.yml

    ??? example "Expected output"
        ```
            everest
            pg
        ```

4. Download the latest release of the [everestctl](https://github.com/percona/percona-everest-cli/releases) command to provision Percona Everest.
5. Rename the downloaded file using the following command and replacing the placeholder `everestctl-darwin-amd64` to match the file downloaded in the previous step: 
    ```sh
    mv everestctl-darwin-amd64 everestctl
    ```
6. Modify the file permissions: 
    ```sh
    chmod +x everestctl
    ```
7. From the installation wizard, provision and register the Kubernetes cluster in Everest using the following command. Everest will search for the kubeconfig file in the `~/.kube/config` path. If your file is located elsewhere, add the `KUBECONFIG` environment variable before running the command.

    !!! note alert alert-primary "Note"
       - If you don't enable monitoring during this provisioning step then you won't be able to enable it from the UI later. Make sure to fill in the monitoring details in the wizard.
       - If you are using a PMM server instance with a self-signed certificate you cannot use HTTPS in the PMM URL endpoint.

    ```sh

>     KUBECONFIG=~/.kube/config; ./everestctl install operators

Reviewer: .kube/config it is not a custom path :) everestctl will use it by default without KUBECONFIG var, I think we can remove it at all
 ```suggestion
   ./everestctl install operators
```

    ```
This will install all needed components in a namespace called `percona-everest`.
       

>     ??? example "Expected output"

Reviewer 4: ```suggestion
    ??? example "Example output"
```

            ```
            ? Everest URL http://127.0.0.1:8080
            ? Choose your Kubernetes Cluster name k3d-dbaas-tilt
            ? Do you want to enable monitoring? Yes
            ? PMM URL Endpoint http://127.0.0.1
            ? Username admin
            ? Password *****
            ? Name for the new monitoring instance my-pmm
            ? Do you want to enable backups? No
            ? What operators do you want to install? MySQL, MongoDB, PostgreSQL
            ```

    Alternatively, you can provision and register the Kubernetes cluster by running the installation in headless mode:
        
    ```

>     KUBECONFIG=~/.kube/config; ./everestctl install operators --backup.enable=false --everest.endpoint=http://127.0.0.1:8080 --operator.mongodb=true --operator.postgresql=true --operator.xtradb-cluster=true --monitoring.enable=true --monitoring.type=pmm --monitoring.new-instance-name=my-pmm --monitoring.pmm.endpoint=http://127.0.0.1 --monitoring.pmm.username=admin --monitoring.pmm.password=admin  --skip-wizard

Reviewer: the same about KUBECONFIG

## Before you start

Percona Everest assists with installing all the necessary operators and required packages, but does not currently help with spinning up a Kubernetes cluster.
Therefore, before installing Everest, ensure that you have a publicly accessible Kubernetes cluster.

We recommend [setting one up on the Amazon Elastic Kubernetes Service (EKS)](../install/eks.md) or on [Google Kubernetes Engine (GKE)](../install/GKE-k8s-cluster), as it may be difficult to currently spin one up on local Kubernetes installations (minikube, kind, k3d, or similar products).

## Prerequisites

Before getting started with Percona Everest, we recommend that you:

1. [Install Docker Engine](hhttps://docs.docker.com/engine/install), which also includes the Docker compose plugin.
2. Install [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) for EKS or the [gcloud CLI](https://cloud.google.com/sdk/docs/install) for GKE.
3. Verify that you have access to the Kubernetes cluster that you want to use with Everest. By default, Everest uses the kubeconfig file available under *~/.kube/config*. Run the following command:
    ```sh 

>     kubectl get nodes`

Reviewer 4: ```suggestion
    kubectl get nodes
```

    ```

    ??? example "Expected output"
        ```
            NAME                                       STATUS   ROLES    AGE   VERSION
             gke-<name>-default-pool-75d48bfc-bx8g   Ready    <none>   11h   v1.26.7-gke.500
             gke-<name>-default-pool-75d48bfc-c2df   Ready    <none>   11h   v1.26.7-gke.500
             gke-<name>-default-pool-75d48bfc-zl7k   Ready    <none>   11h   v1.26.7-gke.500

>          ```

Reviewer 4: This isn't rendered correctly

## Get started

To install and provision Percona Everest:

1. Download the Docker compose file:
```
curl https://raw.githubusercontent.com/percona/percona-everest-backend/release-0.3.0/deploy/quickstart-compose.yml -o quickstart-compose.yml
```

> 2. Initialize the Everest container and its internal PostgreSQL database by executing one of the following commands, each offering distinct access configurations:

Reviewer 5: Why is there only PostgreSQL? When running quick install script, I get all Operators deployed. How to deploy others following these steps?

Reviewer: @nastena1606 I think it is about internal PostgreSQL DB, which Everest uses. It is not about operators at all.

# Install Everest

Percona Everest has two primary components:

 - [Percona CLI](https://github.com/percona/percona-everest-cli), which connects Percona Everest to your Kubernetes cluster.
- Everest application (UI for simplicity) that you can spin up with our quickstart instructions below.

## Before you start

Percona Everest assists with installing all the necessary operators and required packages, but does not currently help with spinning up a Kubernetes cluster.
Therefore, before installing Everest, ensure that you have a publicly accessible Kubernetes cluster.

> We recommend setting one up on the Amazon Elastic Kubernetes Service (EKS) or Google Kubernetes Engine (GKE), as it may be difficult to currently spin one up on local Kubernetes installations (minikube, kind, k3d, or similar products).

Reviewer 6: Since we're copying instructions form oepratros on how to spin up k8s on gke and aws maybe we could link this here?

## Prerequisites
Before getting started with Percona Everest, we recommend that you:

1. Install [Docker](https://docs.docker.com/engine/install/) and docker-compose.
2. Install [Go](https://go.dev/doc/install).
3. Install make.
4. kubeconfig (Everest will use default one from ~/.kube/config)

## Get started
To install and provision Percona Everest:

1. Use the following commands to spin up the backend and frontend:

>  - `wget https://raw.githubusercontent.com/percona/percona-everest-backend/v0.1.0/quickstart.yml`

Reviewer 7: We have v0.2.0 released

Reviewer 7: Do we have any functions to make our versions to update automatically?

- `docker-compose -f quickstart.yml up -d`
2. Download the latest release of the [everestctl](https://github.com/percona/percona-everest-cli/releases) command to provision Percona Everest.
3. Rename the downloaded file as follows: `mv everestctl-darwin-amd64 everestctl`
4. Modify the file permissions: `chmod +x everestctl`

> 5. Rename your kubeconfig file to **config.yaml**, and place it in your present working directory.  (Skip this step if you already have your kubeconfig file in **./kube/config/**).

Reviewer 7: Why do we need to do this?

Author: I thought that if the kubeconfog file is not in the directory used for running Everest installation commands from the terminal, then the next command  won't work. So:

./everestctl install operators --backup.enable=false --everest.endpoint=http://127.0.0.1:8080 --monitoring.enable=false --operator.mongodb=true --operator.postgresql=true --operator.xtradb-cluster=true --skip-wizard -k config.yaml

Let me know if this is not correct and what changes are needed for to this step

6. Run the following command  from the directory where config.yaml is located. This will install all required operators in headless mode:

```sh

> ./everestctl install operators --backup.enable=false --everest.endpoint=http://127.0.0.1:8080 --monitoring.enable=false --operator.mongodb=true --operator.postgresql=true --operator.xtradb-cluster=true --skip-wizard -k config.yaml

Reviewer 6: This might be aproblem if user have kubeconfig in /kube/config/ folder since we're adding flag -k config.yaml it will not work, not sure how to smoothly handle this probably second command (if you have kubeconfg in kube/config run: the same command without last flag

Therefore, before installing Everest, ensure that you have a publicly accessible Kubernetes cluster.

We recommend setting one up on the Amazon Elastic Kubernetes Service (EKS) or Google Kubernetes Engine (GKE), as it may be difficult to currently spin one up on local Kubernetes installations (minikube, kind, k3d, or similar products).

## Prerequisites

Before getting started with Percona Everest, we recommend that you:

1. Install [Docker](https://docs.docker.com/engine/install/) and docker-compose.
2. Install [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html).
3. Download the kubeconfig file from the Kubernetes cluster that you want to use with Everest. By default, Everest uses the file available under *~/.kube/config.*

## Get started

To install and provision Percona Everest:

1. Download the Docker compose file: `curl https://raw.githubusercontent.com/percona/percona-everest-backend/v0.3.0/deploy/quickstart-compose.yml -o quickstart-compose.yml`.
2. Spin up the Everest container and its internal database (PostgreSQL): `docker-compose -f quickstart-compose.yml up -d`.

> 3. Verify if the services started correctly: `docker compose -f quickstart-compose.yml ps --services --filter 'status=running'`. 

Reviewer 4: Maybe we can say this is optional
```suggestion
3. (Optional) Verify if the services started correctly: `docker compose -f quickstart-compose.yml ps --services --filter 'status=running'`. 
```

 ??? example "Expected output"

     everest

>    pg

Reviewer 4: This isn't rendered correctly.

4. Download the latest release of the [everestctl](https://github.com/percona/percona-everest-cli/releases) command to provision Percona Everest.

> 5. Rename the downloaded file as follows: `mv everestctl-darwin-amd64 everestctl`.

Reviewer 4: ```suggestion
5. Rename the downloaded file using the following command and replacing the placeholder `everestctl-darwin-amd64` to match the file downloaded in the previous step: `mv everestctl-darwin-amd64 everestctl`.
```

6. Modify the file permissions: `chmod +x everestctl`.

> 7. From the installation wizard, provision and register the Kubernetes cluster in Everest using the following command. This will create a Percona namespace called `percona-everest`.

Reviewer 4: ```suggestion
7. From the installation wizard, provision and register the Kubernetes cluster in Everest using the following command. This will install all needed components in a namespace called `percona-everest`.
```

   
!!! note alert alert-primary "Note"
    Everest will search for the kubeconfig file in the `~/.kube/config` path. If your file is located elsewhere, add the `KUBECONFIG` environment variable before running the command.

    ```sh 
    KUBECONFIG=~/.kube/config; ./everestctl install operators

>  ??? example "Expected output"

Reviewer 4: This isn't rendered correctly.

    ? Everest URL http://127.0.0.1:8080
    ? Choose your Kubernetes Cluster name k3d-everest-dev
    ? Do you want to enable monitoring? No
    ? Do you want to enable backups? No
    ? What operators do you want to install? MySQL, MongoDB, PostgreSQL

Alternatively, provision and register the Kubernetes cluster in Everest by running the installation in headless mode. 
        
    ```sh
    KUBECONFIG=~/.kube/config; ./everestctl install operators --backup.enable=false --everest.endpoint=http://127.0.0.1:8080 --monitoring.enable=false --operator.mongodb=true --operator.postgresql=true --operator.xtradb-cluster=true --skip-wizard

>     ```

Reviewer 4: This isn't rendered correctly.

    - to limit access to the localhost interface (default): `docker compose -f quickstart-compose.yml up -d`.
    - to expose Everest though a specific interface, use that interface's IP address: `EVEREST_BIND_ADDR=<ip address>; docker compose -f quickstart-compose.yml up -d`.
    - to allow access from any interface on the host machine: `EVEREST_BIND_ADDR=0.0.0.0; docker compose -f quickstart-compose.yml up -d`.
  
3. (Optional) Verify if the services started correctly: `docker compose -f quickstart-compose.yml ps --services --filter 'status=running'`. 
    
    ??? example "Expected output"
        ```
            everest
            pg
        ```

4. Download the latest release of the [everestctl](https://github.com/percona/percona-everest-cli/releases) command to provision Percona Everest.
5. Rename the downloaded file using the following command and replacing the placeholder `everestctl-darwin-amd64` to match the file downloaded in the previous step: `mv everestctl-darwin-amd64 everestctl`.
6. Modify the file permissions: `chmod +x everestctl`.

> 7. From the installation wizard, provision and register the Kubernetes cluster in Everest using the following command: `KUBECONFIG=~/.kube/config; ./everestctl install operators`. 

Reviewer 4: Maybe the command shouldn't be inline and should have a separate code box below the instruction.

> 7. From the installation wizard, provision and register the Kubernetes cluster in Everest using the following command: `KUBECONFIG=~/.kube/config; ./everestctl install operators`. 

Reviewer 4: All commands here are rendered inline. What do you think about having them in a separate code box. This increases readability and add a handy copy button :slightly_smiling_face:

This will install all needed components in a namespace called `percona-everest`.
   
    !!! note alert alert-primary "Note"
        Everest will search for the kubeconfig file in the `~/.kube/config` path. If your file is located elsewhere, add the `KUBECONFIG` environment variable before running the command.

    ??? example "Expected output"

>             ```

Reviewer 4: These ``` are being rendered and shouldn't. Same for others below.
```suggestion
```

# Install Everest

## Setup

Percona Everest has two primary components:

- [Percona CLI](https://github.com/percona/percona-everest-cli), which connects Percona Everest to your Kubernetes cluster.
- Everest application (UI for simplicity), which you can spin up using the instructions below.

### Supported operators

> - Percona Operator for MySQL Based on Percona XtraDB Cluster (PXC) 1.12

Reviewer 3: PXC operator version has changed:
```
kubectl get csv -n percona-everest
NAME                                      DISPLAY                                                      VERSION   REPLACES                                  PHASE
everest-operator.v0.0.21                  Everest operator                                             0.0.21                                              Succeeded
percona-postgresql-operator.v2.2.0        Percona Operator for PostgreSQL                              2.2.0                                               Succeeded
percona-server-mongodb-operator.v1.14.0   Percona Distribution for MongoDB Operator                    1.14.0    percona-server-mongodb-operator.v1.13.1   Succeeded
percona-xtradb-cluster-operator.v1.13.0   Percona Operator for MySQL based on Percona XtraDB Cluster   1.13.0    percona-xtradb-cluster-operator.v1.12.0   Succeeded
```

Reviewer 4: Thanks for spotting this Manish
```suggestion
- Percona Operator for MySQL Based on Percona XtraDB Cluster (PXC) 1.13
```

Reviewer: Please use 1.13.0

2. Install [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) for EKS or the [gcloud CLI](https://cloud.google.com/sdk/docs/install) for GKE.
3. Download the kubeconfig file from the Kubernetes cluster that you want to use with Everest. By default, Everest uses the file available under *~/.kube/config.*

## Get started

To install and provision Percona Everest:

1. Download the Docker compose file:
```
curl https://raw.githubusercontent.com/percona/percona-everest-backend/v0.3.0/deploy/quickstart-compose.yml -o quickstart-compose.yml
```
2. Initialize the Everest container and its internal PostgreSQL database by executing one of the following commands, each offering distinct access configurations:

    - to limit access to the localhost interface (default):
    ```sh
    docker compose -f quickstart-compose.yml up -d
    ```
    - to expose Everest though a specific interface, use that interface's IP address: 
    ```sh
    EVEREST_BIND_ADDR=<ip address>; docker compose -f quickstart-compose.yml up -d
    ```
    - to allow access from any interface on the host machine:
    ```sh

>     EVEREST_BIND_ADDR=0.0.0.0; docker compose -f quickstart-compose.yml up -d

Reviewer 3: This is not secure. I think there was some PR from IT which mentioned that we should not have 0.0.0.0 in quickstart.yml

Reviewer 4: Thanks Manish, we added the `(not recommended)` to make it clear that the users should proceed at their own risk.

Reviewer 4: Actually what the security team complained about was having 0.0.0.0 as the default which we fixed. Now the default is 127.0.0.1 which is safe but we still allow the users to change it if they need to.
